storage/imx6ull-flash/flashdev.c

```diff
@@ -0,0 +1,434 @@
+		/* TODO: check if this actually works */
+		else if (meta->errors[i] == 0xfe) {
+			err = 0;
+			if (i == 0) {
+				blk = (unsigned *)meta->metadata;
+				for (j = 0; j < sizeof(meta->metadata) / sizeof(unsigned); ++j) {
+					err += (sizeof(unsigned) * 8) - __builtin_popcount(*blk + j);
+				}
+
+				if (err > ECC0_BITFLIP_STRENGHT) {
+					return -EBADMSG;
+				}
+
+				memset(blk, 0xff, sizeof(meta->metadata));
+				max_bitflip = max(max_bitflip, err);
+			}
+			else {
+				blk = data + ((i - 1) * 512);
+				for (j = 0; j < 128; j++) {
+					err += (sizeof(unsigned) * 8) - __builtin_popcount(*(blk + j));
+				}
+
+				if (err > ECCN_BITFLIP_STRENGHT) {
+					return -EBADMSG;
+				}
+
+				memset(blk, 0xff, 512);
+				max_bitflip = max(max_bitflip, err);
+			}
+		}
+		/* correctable errors */
+		else {
+			max_bitflip = max(max_bitflip, (int)meta->errors[i]);
+		}
+	}
+
+	return max_bitflip >= ECC_BITFLIP_THRESHOLD ? -EUCLEAN : status;
+}
+
+
+/* MTD interface */
+
+static int flashmtd_erase(struct _storage_t *strg, off_t offs, size_t size)
+{
+	int res;
+	blkcnt_t pageID;
+	blkcnt_t eb, startBlock, endBlock, erased = 0; /* values in erase block units */
+	unsigned int pagesPerBlock;
+
+	if ((offs % strg->dev->mtd->erasesz) != 0 || (size % strg->dev->mtd->erasesz) != 0) {
+		return -EINVAL;
+	}
+
+	startBlock = offs / strg->dev->mtd->erasesz;
+	endBlock = (offs + size) / strg->dev->mtd->erasesz;
+	pagesPerBlock = strg->dev->mtd->erasesz / strg->dev->mtd->writesz;
+
+	mutexLock(strg->dev->ctx->lock);
+	for (eb = startBlock; eb < endBlock; ++eb) {
+		pageID = eb * pagesPerBlock;
+
+		res = flashdrv_isbad(strg->dev->ctx->dma, pageID);
+		if (res != 0) {
+			printf("flashmtd_erase: skipping bad block: %lld\n", eb);
+			continue;
+		}
+
+		res = flashdrv_erase(strg->dev->ctx->dma, pageID);
+		if (res < 0) {
+			printf("flashmtd_erase: error while erasing block: %lld - marking as badblock", eb);
+			res = flashdrv_markbad(strg->dev->ctx->dma, pageID);
+			if (res < 0) {
+				mutexUnlock(strg->dev->ctx->lock);
+				return res; /* can't mark as badblock, urecoverable error */
+			}
+		}
+		else {
+			++erased;
+		}
+	}
+	mutexUnlock(strg->dev->ctx->lock);
+
+	return erased;
+}
+
+
+static ssize_t flashmtd_read(struct _storage_t *strg, off_t offs, void *data, size_t len)
+{
+	int res, err = 0;
+	size_t tempsz = 0, chunksz;
+
+	mutexLock(strg->dev->ctx->lock);
+	if (offs % strg->dev->mtd->writesz) {
+		res = flashdrv_read(strg->dev->ctx->dma, offs / strg->dev->mtd->writesz, strg->dev->ctx->databuf, strg->dev->ctx->metabuf);
+		err = _flashmtd_errRead(res, strg->dev->ctx->metabuf, err, strg->dev->ctx->databuf);
+		if (err < 0) {
+			mutexUnlock(strg->dev->ctx->lock);
+			return err;
+		}
+
+		chunksz = (strg->dev->mtd->writesz - (offs % strg->dev->mtd->writesz));
+		if (chunksz > len) {
+			chunksz = len;
+		}
+
+		memcpy(data, strg->dev->ctx->databuf + (offs % strg->dev->mtd->writesz), chunksz);
```

Comment:
`void *` arithmetic on `databuf` buffer.
Maybe change buffers defition to `unsigned char *`?

Reply:
Thanks. I have also changed in the rest of the code.